internal/yaml.go

```diff
@@ -0,0 +1,157 @@
+// Copyright 2021 FerretDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal
+
+import (
+	"fmt"
+
+	"golang.org/x/exp/maps"
+)
+
+// Stats represent the expected/actual amount of
+// failed, skipped and passed tests.
+type Stats struct {
+	UnexpectedRest int `yaml:"unexpected_rest"`
+	UnexpectedFail int `yaml:"unexpected_fail"`
+	UnexpectedSkip int `yaml:"unexpected_skip"`
+	UnexpectedPass int `yaml:"unexpected_pass"`
+	ExpectedFail   int `yaml:"expected_fail"`
+	ExpectedSkip   int `yaml:"expected_skip"`
+	ExpectedPass   int `yaml:"expected_pass"`
+}
+
+// ConfigFile is a yaml tests representation of the Config struct.
+//
+// It is used only to fetch data from file. To get any of
+// the dance configuration data it should be converted to
+// Config struct with Convert() function.
+//
+//nolint:govet // we don't care about alignment there
+type ConfigFile struct { // ConfigFile -> Config
+	Runner  string     `yaml:"runner"`
+	Dir     string     `yaml:"dir"`
+	Args    []string   `yaml:"args"`
+	Results resultList `yaml:"results"`
+}
+
+// resultList is a yaml representation of the Results struct.
+type resultList struct { // resultList -> ResultList
+	Common   *testsConfig `yaml:"common"`
+	FerretDB *testsConfig `yaml:"ferretdb"`
+	MongoDB  *testsConfig `yaml:"mongodb"`
+}
+
+// testsConfig is a yaml representation of the TestsConfig struct.
+// It differs from it by using "any" type to be able to parse maps (i.e. "- output_regex: ...").
+//
+// To gain a data the struct should be first converted to TestsConfig with testsConfig.Convert() function.
+type testsConfig struct { // testsConfig -> TestsConfig
+	Default status `yaml:"default"`
+	Stats   *Stats `yaml:"stats"`
+	Pass    []any  `yaml:"pass"`
+	Skip    []any  `yaml:"skip"`
+	Fail    []any  `yaml:"fail"`
+}
```

Comment:
PR ready for review.
As you can see here we have 3 structs which are only used to fetch yaml data. Every of them has it's "internal" representation.
I can see here a little name inconsistency. We've discussed it with @seeforschauer and my point is that:
- The `Config` yaml representation adds a "File" part at the end.
- The `resultList` and `testsConfig` representation is just exported (which maybe would not be a bad idea if the structs were in the seperate package, but they're not).

I think that current state is a little bit misleading (especially for somone new to the codebase).

@AlekSi please let me know what do you think and maybe if you have some suggestions in that case.

Reply:
Yep, here something can be done better probably.

My point is that because we made an analogy for corresponded structures - it's easy to understand.
```go
// resultList -> ResultList
// testsConfig -> TestsConfig
```

It seems too little code to extract it into a separate package and make it unexported. 
And also probably will overcomplicate things.

Reply:
I think having types `resultList` and `ResultList` that differ only in case is confusing, especially given that both of them are not used outside of the package. Suffix would be better, in my opinion. For example, `results` (that's not even a list) and `resultsYAML`.

Reply:
Alternatively, `Results` and `ResultsYAML`. Both are exported, but they also were exported before this PR, so we just don't change it now. Maybe letter we will split them into a few packages

Reply:
> Alternatively, `Results` and `ResultsYAML`. Both are exported, but they also were exported before this PR, so we just don't change it now. Maybe letter we will split them into a few packages

Vote for this.

Reply:
Okay, but should we use "yaml" or maybe something less language specific like "file"?

Reply:
Well, that's not a whole file, only part of it. It is also has only `yaml` tags, not `json` or `toml` tags, and we don't support anything other than YAML. So I propose to ~[paint this bikeshed green](https://bikeshed.org)~ call it `ResultsYAML`.